domain/modeldefaults/state/state.go

```diff
@@ -230,10 +235,276 @@ WHERE m.uuid = ?
+`, model, region)
+	if err != nil {
+		return "", "", errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, query, model).Get(&region)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return interrors.Errorf("model %q not found%w", uuid, errors.Hide(modelerrors.NotFound))
+		}
+		return errors.Trace(err)
+	})
+	return cloud.UUID(region.CloudUUID), region.Name, errors.Annotatef(err, "getting cloud details for model %q", uuid)
+}
+
+// GetCloudUUID returns the cloud UUID and region for the given cloud name.
+// If the cloud is not found, an error specifying [clouderrors.NotFound] is returned.
+func (s *State) GetCloudUUID(ctx context.Context, cloudName string) (cloud.UUID, error) {
+	db, err := s.DB()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	cld := dbCloud{Name: cloudName}
+	query, err := s.Prepare(`
+SELECT &dbCloud.uuid
+FROM cloud c
+WHERE c.name = $dbCloud.name
+`, cld)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, query, cld).Get(&cld)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return interrors.Errorf("cloud %q not found%w", cloudName, errors.Hide(clouderrors.NotFound))
+		}
+		return errors.Trace(err)
+	})
+	return cloud.UUID(cld.UUID), errors.Annotatef(err, "getting cloud UUID for %q", cloudName)
+}
+
+// UpdateCloudDefaults is responsible for updating default config values for a
+// cloud. This function will allow the addition and updating of attributes.
+// If the cloud doesn't exist, an error satisfying [clouderrors.NotFound]
+// is returned.
+func (st *State) UpdateCloudDefaults(
+	ctx context.Context,
+	cloudUUID cloud.UUID,
+	updateAttrs map[string]string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	cld := dbCloud{UUID: cloudUUID.String()}
+
+	upsertStmt, err := sqlair.Prepare(`
+INSERT INTO cloud_defaults (cloud_uuid, key, value)
+VALUES ($cloudDefaultValue.*)
+ON CONFLICT(cloud_uuid, key) DO UPDATE
+    SET value = excluded.value
+    WHERE cloud_uuid = excluded.cloud_uuid
+    AND key = excluded.key;
+`, cloudDefaultValue{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		for k, v := range updateAttrs {
+			err := tx.Query(ctx, upsertStmt, cloudDefaultValue{UUID: cld.UUID, Key: k, Value: v}).Run()
+			// The cloud UUID has previously been checked. This allows us to avoid having to use RunAtomic.
+			if database.IsErrConstraintForeignKey(err) {
+				return interrors.Errorf("cloud %q not found%w", cloudUUID, errors.Hide(clouderrors.NotFound))
+			} else if err != nil {
+				return interrors.Errorf("updating cloud default keys %q: %w", cloudUUID, err)
+			}
+		}
+		return nil
+	})
+}
+
+// DeleteCloudDefaults deletes the specified cloud default
+// config values if they exist.
+func (st *State) DeleteCloudDefaults(
```

Comment:
Should this have a pre-flight check to make sure the cloud exists? If so I think we should remove the len == 0 check.